projects/client/RabbitMQ.Client/src/client/impl/SocketFrameHandler.cs

```diff
@@ -1,321 +1,304 @@
+            get { return ((IPEndPoint)LocalEndPoint).Port; }
+        }
+
+        public int ReadTimeout
+        {
+            set
+            {
+                try
+                {
+                    if (m_socket.Connected)
+                    {
+                        m_socket.ReceiveTimeout = value;
+                    }
+                }
+                catch (SocketException)
+                {
+                    // means that the socket is already closed
+                }
+            }
+        }
+
+        public int WriteTimeout
+        {
+            set
+            {
+                m_writeableStateTimeout = value;
+                m_socket.Client.SendTimeout = value;
+            }
+        }
+
+        public void Close()
+        {
+            lock (_semaphore)
+            {
+                if (!_closed)
+                {
+                    try
+                    {
+                        m_socket.Close();
+                    }
+                    catch (Exception)
+                    {
+                        // ignore, we are closing anyway
+                    }
+                    finally
+                    {
+                        _closed = true;
+                    }
+                }
+            }
+        }
+
+        public ReadFrame ReadFrame()
+        {
+            return RabbitMQ.Client.Impl.ReadFrame.ReadFrom(m_reader);
+        }
+
+        private static readonly byte[] amqp = Encoding.ASCII.GetBytes("AMQP");
+        public void SendHeader()
+        {
+            var ms = new MemoryStream();
+            var nbw = new NetworkBinaryWriter(ms);
+            nbw.Write(amqp);
+            byte one = (byte)1;
+            if (Endpoint.Protocol.Revision != 0)
+            {
+                nbw.Write((byte)0);
+                nbw.Write((byte)Endpoint.Protocol.MajorVersion);
+                nbw.Write((byte)Endpoint.Protocol.MinorVersion);
+                nbw.Write((byte)Endpoint.Protocol.Revision);
+            }
+            else
+            {
+                nbw.Write(one);
+                nbw.Write(one);
+                nbw.Write((byte)Endpoint.Protocol.MajorVersion);
+                nbw.Write((byte)Endpoint.Protocol.MinorVersion);
+            }
+            m_writer.Write(ms.ToArray());
+        }
+
+        public void WriteFrame(WriteFrame frame)
+        {
+            var ms = new MemoryStream();
+            var nbw = new NetworkBinaryWriter(ms);
+            frame.WriteTo(nbw);
+            m_socket.Client.Poll(m_writeableStateTimeout, SelectMode.SelectWrite);
+            m_writer.Write(ms.ToArray());
+        }
+
+        public void WriteFrameSet(IList<WriteFrame> frames)
+        {
+            var ms = new MemoryStream();
+            var nbw = new NetworkBinaryWriter(ms);
+            foreach (var f in frames) f.WriteTo(nbw);
+            m_socket.Client.Poll(m_writeableStateTimeout, SelectMode.SelectWrite);
+            m_writer.Write(ms.ToArray());

```

Comment:
Entire frame is written as one unit to the stream, thus removing the need to worry about frame interleaving.